src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpReceiver.php

```diff
@@ -61,24 +62,34 @@ public function receive(callable $handler): void
                 ]));
 
                 $this->connection->ack($AMQPEnvelope);
-            } catch (RejectMessageExceptionInterface $e) {
+            } catch (RecoverableMessageExceptionInterface $e) {
                 try {
-                    $this->connection->reject($AMQPEnvelope);
+                    $this->connection->nack($AMQPEnvelope, AMQP_REQUEUE);
+                } catch (\AMQPException $exception) {
+                    throw new TransportException($exception->getMessage(), 0, $exception);
+                }
+            } catch (UnrecoverableMessageExceptionInterface $e) {
+                try {
+                    $this->connection->nack($AMQPEnvelope, AMQP_NOPARAM);
                 } catch (\AMQPException $exception) {
                     throw new TransportException($exception->getMessage(), 0, $exception);
                 }
-
-                throw $e;
             } catch (\AMQPException $e) {
                 throw new TransportException($e->getMessage(), 0, $e);
             } catch (\Throwable $e) {
+                $connectionCredentials = $this->connection->getConnectionCredentials() + [
+                        'consume_fatal' => true,
+                        'consume_requeue' => false,
+                ];
+                $flag = $connectionCredentials['consume_requeue'] ? AMQP_REQUEUE : AMQP_NOPARAM;
                 try {
-                    $this->connection->nack($AMQPEnvelope, AMQP_REQUEUE);
+                    $this->connection->nack($AMQPEnvelope, $flag);
                 } catch (\AMQPException $exception) {
                     throw new TransportException($exception->getMessage(), 0, $exception);
                 }
-
-                throw $e;
+                if ($connectionCredentials['consume_fatal']) {
+                    throw $e;
+                }
```

Comment:
I'm very interested in this part, and these two new options. I'd like to make sure we're thinking about a solution that could be generically applied to other transports and support a "retry" ability. Here's the behavior I'd like to see, which is built off of your ideas:

1) If `RecoverableMessageExceptionInterface` is thrown, we re-queue
2) If `UnrecoverableMessageExceptionInterface` is thrown, we discard
3) If there is an `AMQPException` at any point, probably do nothing - because the message was never `ack`'ed, so it should still be on the queue? But, it should not kill the worker process (but that's another PR)
3) If *any* other `Throwable` occurs, I think we should requeue always and *never* throw the exception. However, a few caveats:
 * it should only requeue the message a certain number of times before discarding it (that's the retry stuff)
 * once it hits the retry max, the exception is still not thrown, but perhaps an event is dispatched to handle this.

Morever, I think one of your goals was to standardize a few things (even though we don't have multiple transports yet), which I love! Which is why I'm wondering if we should move some of this logic to the `Worker` class - #30557 is related. I'm proposing that this method should not handle any exceptions or requeue - it should just handle the message and try to `ack` it on success. We would move the exception-handling logic into `Worker` (so we can standardize behavior) and add a few new methods to `ReceiverInterface`, like `requeu()` or `discard()`. Worker would call this under the appropriate situations.

WDYT?

Reply:
Yes, could be a goog idea to move this into the worker ! +1 :)

Reply:
For your point 4, I wanted to keep the compatibility. As you wish :)